openedx/core/djangoapps/discussions/docs/decisions/0006-discussions-api-fetch-inactive-providers.rst

```diff
@@ -0,0 +1,82 @@
+Allow fetching configuration for inactive providers
+===================================================
+
+
+Status
+------
+
+Proposal
+
+
+Context
+-------
+
+In ADR 003 an API was proposed for discussion configuration. This proposed a
+single API that returned information about the active provide and listed other
+available providers.
+
+The API was designed to only return information about the current active
+provider however in practise this creates an issue when switching providers.
+
+The troublesome scenario is as follows:
+
+- A course author configures the legacy provider
+- They switch to another provider
+- Now they switch back to the legacy provider
+- They have to set up the legacy provider from scratch again
+
+Since we can't fetch configuration for an inactive provider, the wizard cannot
+use their existing configuration for the legacy provider.
```

Comment:
Why does the wizard run at all in this case? I would assume that when "they switch back to the legacy provider", it could just restore all their previous settings without any "wizard" or user input required at all. After all, if this API proposes exposing the old configuration data, then it must still be kept around somewhere, right?

In other words, why is step 4 `They have to set up the legacy provider from scratch again` required? I think that's missing from this ADR.

Reply:
When you click on the discussions tile in the course authoring MFE, you are presented with a list of supported discussion providers. Here you can just click next to continue with the existing provider, and then you see a page with configuration for the provider. When you save this configuration, all the settings are applied. 

If, in the first step, you pick a different provider (for example, you're changing from Piazza to the `legacy` provider) the next step in the wizard should show the existing configuration for the legacy provider, however the configuration it currently has loaded is for the active (Piazza) provider. So despite having the configuration, it cannot show it, since the frontend doesn't have this information. This change will allow the API to fetch this data from the backend so that it can show the actual saved settings.

Reply:
I see. Well I would suggest you add to the ADR an alternative which is : "keep the API the same but modify the UI so that when selecting a different provider, it displays a message like 'You have changed discussion providers. Before you can configure this provider, you'll need to click Save Changes.'"

You may decide not to go with that approach but I think it's worth considering (in the ADR text) since it's so much easier and this is an edge case anyways.

Reply:
@bradenmacdonald Roughly the approach you are mentioning is already considered and part of the ADR in the "Alternatives Considered" section, i.e. that clicking next after changing the provider would first update the provider using one API call, fetch the new settings, and then continue to the next page.

Reply:
@xitij2000 Ah, that wasn't clear to me. I thought that what that section was saying was:

"An alternative is: if the frontend wishes to retrieve the configuration for an inactive provider, it can make an API requests to temporarily change the active provider, fetch its configuration, and then change the provider back."

Which yes that does sound "disruptive"

If it's just saying "users are required to save changes before configuring a provider in detail", I don't think that's particularly disruptive.

In any case, it sounds like you have a good plan here, so I don't want to de-rail everything. I am mostly making these comments in hopes that the language in the ADR can be a bit clearer because those things weren't clear to me after reading it.

Reply:
> If it's just saying "users are required to save changes before configuring a provider in detail", I don't think that's particularly disruptive.

Yes, the disruptive aspect is that changing the provider leads to a full publish cycle that would happen twice, once for the provider change and once for the settings change. If we can just fetch the settings without activating the provider if possible, that's better. 

I am going to re-write parts of the ADR that are under-explained or confusing.

Reply:
I've updated the ADR to be clearer.